Products/Jobber/model.py

```diff
@@ -0,0 +1,237 @@
+##############################################################################
+#
+# Copyright (C) Zenoss, Inc. 2019, all rights reserved.
+#
+# This content is made available according to terms specified in
+# License.zenoss under the directory where your Zenoss product is installed.
+#
+##############################################################################
+
+from __future__ import absolute_import
+
+import logging
+import os
+import time
+
+from celery import states
+from celery.contrib.abortable import ABORTED
+from zope.component import getUtility
+from zope.interface import implementer
+
+from Products.Zuul.interfaces import IMarshaller, IMarshallable
+
+from .config import ZenJobs
+from .interfaces import IJobStore, IJobRecord
+from .storage import Fields
+from .utils.log import inject_logger
+from .zenjobs import app
+
+mlog = logging.getLogger("zen.zenjobs.model")
+
+
+@implementer(IJobRecord, IMarshallable)
+class JobRecord(object):
+    """Zenoss-centric record of a job submitted to ZenJobs."""
+
+    __slots__ = tuple(key for key in Fields.keys())
+
+    @classmethod
+    def make(cls, data):
+        if not (data.viewkeys() <= Fields.viewkeys()):
+            bad = data.viewkeys() ^ Fields.viewkeys()
+            raise AttributeError(
+                "Jobrecord does not have attribute%s %s" % (
+                    "" if len(bad) == 1 else "s",
+                    ", ".join("'%s'" % v for v in bad),
+                ),
+            )
+        record = cls()
+        for k, v in data.viewitems():
+            setattr(record, k, v)
+        return record
+
+    def __getattr__(self, name):
+        # Clever hack for backward compatibility.
+        # Users of JobRecord added arbitrary attributes.
+        if name not in self.__slots__:
+            details = getattr(self, "details", {}) or {}
```

Comment:
I'm assuming this is where the zenoss-specific properties live in the JobRecord, but i'm having trouble figuring out where it gets stored and populated.

Reply:
Instances of the new `JobRecord` class prevents the setting of attributes.  The original code did not do this, but any attempt to set an attribute on a `JobRecord` object was a bug because the catalog would not be updated.

In both the new and old code, the `update` method of the `JobManager` object is the correct way to set attributes on job records.  Jobs have a `setProperties` method that will add or update entries to the `details` field.  The `setProperties` method saves the updated `details` field using `JobManager`'s `update` method.

Properties (i.e. details) can also be set when the job is created using `JobManager`'s `addJob` method.  The `JobRecord` data is saved after the ZODB transaction has completed its commit.  When the transaction's commit successfully completes, the task is sent and the "before_task_publish" event is fired right before the task is written to the queue. All "before_task_publish" event listeners are run.  The `save_jobrecord` function found in Products/Jobber/model.py is the listener configured to handle "before_task_publish" events. The `save_jobrecord` function saves the job record to Redis.

Other than `JobManager`'s `update` method, job record updates are handled via event handlers.  There are events, or "signals" as Celery calls them, for most state changes in Celery's task and worker lifecycles.  The event handlers are registered using ZCML.  The **configure.zcml** file contains the "before_task_publish" registration and **signals.zcml** contains all the other registrations.  Only the Celery worker instance (i.e. zenjobs service) loads the **signals.zcml** file.